### PR TITLE
uses designated thread-pools for account db parallel ops

### PR DESCRIPTION
**Pull request by behzadnouri**
#### Problem
accounts-db ops running in global thread-pool taking excessive resources.

#### Summary of Changes
use designated thread-pools from accounts-db.

### CODE REVIEW

File: runtime/src/accounts_db.rs [hunk 1]
```diff
@@ -6156,32 +6162,34 @@ impl AccountsDb {
...
-                        info,
-                        &mut reclaims,
-                        previous_slot_entry_was_cached,
-                    );
-                });
-                reclaims
-            })
-            .flatten()
-            .collect::<Vec<_>>()
+        self.thread_pool.install(|| {
```
Comment (jeffwashington) [inline runtime/src/accounts_db.rs:6165]: This function can be called from both fg (tx processing, bank freeze) and bg processes (clean, shrink). I have a comment above where when I tried to put a thread pool on here we got deadlocks. Perhaps I was trying to use the bg thread pool. I specifically implemented this function to use a controlled number of threads max and avoided a thread pool.

File: runtime/src/accounts_db.rs [hunk 2]
```diff
@@ -1804,17 +1804,23 @@ impl AccountsDb {
...
-            .map(|pubkeys: &[Pubkey]| {
-                let mut reclaims = Vec::new();
-                for pubkey in pubkeys {
-                    self.accounts_index
-                        .clean_rooted_entries(pubkey, &mut reclaims, max_clean_root);
-                }
-                reclaims
-            });
-        let reclaims: Vec<_> = reclaim_vecs.flatten().collect();
+        let reclaims: Vec<_> = self.thread_pool_clean.install(|| {
```
Comment (jeffwashington) [inline runtime/src/accounts_db.rs:1807]: this one looks good.

File: runtime/src/accounts_db.rs (hunk 1 above)
Comment (behzadnouri) [reply runtime/src/accounts_db.rs]: We can debug deadlocks if they are reproducible but as it stands, this is still taking a lot of resources; and we have a performance regression on current master code which I am trying to debug.
I see the function is invoked from two paths:
* `store_accounts_frozen` which I can only see it be invoked from cleanup and bg process.
* `store_accounts_unfrozen` which only seems to be invoked from replay and other fg processes.

If this is true then we can pass the thread-pool explicitly to `update_index` and pick the right thread-pool in `store_accounts_frozen` and `store_accounts_unfrozen`. Last commit on the PR is implementing this.

File: runtime/src/accounts_db.rs [hunk 3]
```diff
@@ -6156,32 +6163,34 @@ impl AccountsDb {
         let len = std::cmp::min(accounts.len(), infos.len());
```
Comment (jeffwashington) [inline runtime/src/accounts_db.rs:6163]: Need to adjust or remove the comment right above here.
I imagine I was using the bg thread pool and that was causing a deadlock. We can ignore all of that as you have done by passing a thread pool and making sure we pass the right thread pool.

File: runtime/src/accounts_db.rs [hunk 4]
```diff
@@ -6620,6 +6629,7 @@ impl AccountsDb {
         }
     }
 
+    // TODO: confirm this is only called from replay and fg processes.
```
Comment (jeffwashington) [inline runtime/src/accounts_db.rs:6632]: These todos seem probably true today. I agree we should confirm this. I imagine we'd find deadlocks if something about the code changes and we encounter thread pool crossing code at some point in the future.

**Comment by jeffwashington** [comment]
@behzadnouri see #24692 also. This is another solution and some metrics.

**Comment by jeffwashington** [comment]
> If using thread-pools here is unnecessary, then yeah for sure better to remove them all together.

this isn't resolved yet and won't work well without thread pools as # accounts grow. I think for now we still want the parallelism.

File: runtime/src/accounts_db.rs (hunk 3 above)
Comment (behzadnouri) [reply runtime/src/accounts_db.rs:6166]: update the comments.
Do we still want that manual batching and `chunk_size`, etc?

File: runtime/src/accounts_db.rs (hunk 4 above)
Comment (behzadnouri) [reply runtime/src/accounts_db.rs]: Besides a mainnet node which was running with this code for a while, I have not done any other form of testing yet to see if this deadlocks or not.

What was the setup that you were observing dead-locks?
Meaning how long did it take until you it dead-locked? and, was it just a normal validator running on mainnet, or was it in a test setup?

Comment (jeffwashington) [reply runtime/src/accounts_db.rs]: It was likely a mnb validator. If it isn't deadlocking using these thread pools in your tests, then we most likely aren't aggravating that or something else has changed somewhere else in the system. I'm confident I was trying to use the clean thread pool.

**Review by jeffwashington — APPROVED** [review]
lgtm

**Pull request merged**